charts/scalardb-cluster/Chart.lock

```diff
@@ -1,6 +1,6 @@
 dependencies:
 - name: envoy
-  repository: https://scalar-labs.github.io/helm-charts
-  version: 2.2.0
-digest: sha256:2400613e162d0acb575e173d8623a67281e131d0e00b0da7925177e793c13643
-generated: "2022-09-14T12:34:38.629511121+09:00"
+  repository: file://../envoy/
```

Comment:
Is using `file://` OK, here? (I just don't know about it 🙇 )
If it is OK, I'm just wondering why it wasn't like this.

Reply:
@feeblefakie  
Thank you for your review!

> Is using `file://` OK, here? (I just don't know about it 🙇 )

Yes. Using `file://` is expected.

> If it is OK, I'm just wondering why it wasn't like this.

Because we couldn't access the SNAPSHOT version in the `main` branch from the topic branch (`add-scalardb-cluster-chart`) that we implement the ScalarDB Cluster chart. But, I think we could avoid this issue by merging the `main` branch with the topic branch. Sorry, it was my mistake.

---

I explain in more detail to answer your question as follows.

### 1. How to use dependency chart in `main` branch (i.e., using SNAPSHOT versions)

When we use stable versions, we can specify the Scalar Helm Repository to resolve dependencies in each chart. This is because, the dependency chart (i.e., Scalar Envoy) is published in the Scalar Helm Repository.

* Stable versions
  ```console
  [ScalarDB Cluster]---(depends on)--->[Scalar Envoy (published in Helm Repository)]
  ```

However, we don't publish the SNAPSHOT version chart in the Scalar Helm Repository since it is not ready for production (i.e., it is not for users). Basically, we assume developers of Scalar products (including Scalar Helm Chart) use the SNAPSHOT version charts. So, we cannot specify the dependency via Scalar Helm Repository.

* SNAPSHOT versions
  ```console
                                        We cannot specify this dependency by using Scalar Helm Repository since it is not published
                                       /
  [ScalarDB Cluster SNAPSHOT]---(depends on)---X--->[Scalar Envoy SNAPSHOT (not published in Helm Repository)]
  ```

Also, for Scalar products (including Scalar Helm Chart itself) developers, we assume that they use the SNAPSHOT version charts as follows.

1. git clone https://github.com/scalar-labs/helm-charts.git
1. cd helm-charts/
1. helm dependency build ./charts/<CHART_NAME>/
1. helm install <RELEASE_NAME> ./charts/<CHART_NAME>/ -f <CUSTOM_VALUES_FILE>

Note: The CI deploys each Scalar product by the same way (get charts by git clone).

In this case, we have to specify the dependency based on the file path. So, we have to specify the `file://` to specify the dependency of SNAPSHOT version charts.

For example, we have to specify `file://../envoy/` based on the directory structure (relative path) if we specify the Envoy chart from ScalarDB Cluster chart.

* SNAPSHOT versions
  ```console
  [ScalarDB Cluster SNAPSHOT]---(specify dependency based on the file path)--->[Scalar Envoy SNAPSHOT]
  ```
  ```console
  ├── charts
  │   ├── envoy
  │   │   ├── Chart.yaml
  │   │   ├── README.md
  │   │   ├── ...
  │   │    ...
  │   │
  │   ├── scalardb-cluster
  │   │   ├── Chart.lock
  │   │   ├── Chart.yaml
  │   │   ├── ...
  │   │    ...
  ```

### 2. Why we didn't use SNAPSHOT Envoy chart when we develop ScalarDB Cluster chart in `add-scalardb-cluster-chart` branch

I developed the `ScalarDB Cluster chart` and `Using SNAPSHOT version` in this repository in the following order.

1. Create `add-scalardb-cluster-chart` branch
1. Update `main` branch to use SNAPSHOT versions

So, the topic branch `add-scalardb-cluster-chart` didn't include the SNAPSHOT version Envoy. So, I specified the stable version of Envoy in this topic branch when I developed the ScalarDB Cluster chart.

However, we could merge SNAPSHOT version updates into the topic branch `add-scalardb-cluster-chart` after SNAPSHOT version updates are applied to the `main` branch. Also, in this way, we could test the ScalarDB Cluster chart by using the SNAPSHOT Envoy chart before we merge `add-scalardb-cluster-chart` into `main` in PR #171.

But, I didn't do so. I overlooked that. I noticed it when I release v1.0.0.
This is simply my mistake, sorry for that...

Reply:
Thank you!
I kind of got it :)